### PR TITLE
chore: enforce buyAmountAfterFeesCryptoBaseUnit at swapper level

### PR DESCRIPTION
**Pull request by 0xApotheosis**
## Description

- Enforce `buyAmountAfterFeesCryptoBaseUnit` at the swapper level to avoid the need for client-side logic, and improve our swapper SDK/API interface.

- Fixes some bugs where the `buyAmountBeforeFeesCryptoBaseUnit` was incorrect.

- Slippage in the UI now shows the expected receive amount after slippage (what a great idea), rather than the same as the "you get" amount. The "you get" amount now shows the expected amount, rather than the worst case minimum reciived.

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [x] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

Closes https://github.com/shapeshift/web/issues/5171

## Risk

High risk to:

- the amounts we show the user, before and after fees
- the amounts we are giving to the quote and trade execution requests

Test accordingly 🙏

## Testing

Get quotes for each swapper and compare the amounts we show against the requset/response payloads.

Ensure the amount that is recieved when executing transactions is in line with the amount we show to the user.

### Engineering

☝️

### Operations

☝️

## Screenshots (if applicable)

N/A

### CODE REVIEW

**Comment by 0xApotheosis** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #5175** <a href="https://app.graphite.dev/github/pr/shapeshift/web/5175" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/5175?utm_source=stack-comment).

File: src/lib/swapper/swappers/LifiSwapper/getTradeQuote/getTradeQuote.ts [hunk 1]
```diff
@@ -152,7 +152,32 @@ export async function getTradeQuote(
...
+            ).times(estimateRate)
+            const buySideProtocolFeeCryptoBaseUnit = bnOrZero(
+              buyAssetProtocolFee?.amountCryptoBaseUnit,
+            )
+
+            const buyAmountAfterFeesCryptoBaseUnit = bnOrZero(
+              selectedLifiRoute.toAmount,
+            ).toPrecision()
+
+            // TODO: Add buySideNetworkFeeCryptoBaseUnit when we implement multihop
```
Comment (gomesalexandre) [inline src/lib/swapper/swappers/LifiSwapper/getTradeQuote/getTradeQuote.ts:175]: Read my mind there ser!

File: src/lib/swapper/swappers/ZrxSwapper/getZrxTradeQuote/getZrxTradeQuote.ts [hunk 2]
```diff
@@ -67,10 +67,16 @@ export async function getZrxTradeQuote(
   if (maybeZrxPriceResponse.isErr()) return Err(maybeZrxPriceResponse.unwrapErr())
   const { data } = maybeZrxPriceResponse.unwrap()
 
-  const useSellAmount = !!sellAmountIncludingProtocolFeesCryptoBaseUnit
-  const rate = useSellAmount ? data.price : bn(1).div(data.price).toString()
+  const {
+    buyAmount: buyAmountAfterFeesCryptoBaseUnit,
+    grossBuyAmount: buyAmountBeforeFeesCryptoBaseUnit,
```
Comment (gomesalexandre) [inline src/lib/swapper/swappers/ZrxSwapper/getZrxTradeQuote/getZrxTradeQuote.ts:72]: 🧠

**Review by gomesalexandre — COMMENTED** [review]
Very first pass, plumbing looks good - untested yet

**Review by gomesalexandre — APPROVED** [review]
Second pass, confirmed this does what it says on the box from a functional perspective 🏅 :

### THOR

- Expected amount vs. request/response payload

Screenshot 2023-08-28 at 20 51 40
Screenshot 2023-08-28 at 20 51 30


- Actual amount received

image

### 0x 

- Expected amount vs. request/response

Screenshot 2023-08-28 at 21 05 43

- Actual received amount

image

### Same chain Li.Fi

- Expected amount vs. request/response

image
image

- Actual received amount

image

### Cross-Chain Li.Fi

- Expected amount vs. request/response

image
image

- Actual received amount

image

### CoW

- Expected amount vs. request/response

image
image

- Actual received amount

image

File: src/state/apis/swappers/helpers/getInputOutputRatioFromQuote.ts [hunk 3]
```diff
@@ -139,21 +90,16 @@ const _convertCryptoBaseUnitToUsdPrecision = (
...
-  const { buyAsset } = quote.steps[quote.steps.length - 1]
-
-  const { netReceiveAmountCryptoBaseUnit, buySideNetworkFeeCryptoBaseUnit } =
-    _getReceiveSideAmountsCryptoBaseUnit({
-      quote,
-      swapperName,
-    })
+  const lastStep = quote.steps[quote.steps.length - 1]
+  const { buyAsset, buyAmountAfterFeesCryptoBaseUnit: netReceiveAmountCryptoBaseUnit } = lastStep
+  const buySideNetworkFeeCryptoBaseUnit = bn(0)
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/state/apis/swappers/helpers/getInputOutputRatioFromQuote.ts:103]: Good call, added TODO in https://github.com/shapeshift/web/pull/5175/commits/3af4822b2fdadf0fde752cc19c2178a67da133f9!

**Pull request merged**